### ISSUE TITLE
Bug: Cancelled delayed transition on transition in parallel states (v5)

### ISSUE THREAD

**Issue by rlaffers** [statelyai/xstate#4360]
### Description

If a transition into a parallel state is executed, the pending delayed transition in the current state node is cancelled.

```javascript
Waiting: { 
  after: {
    2000: "Fetching"  // a delayed transition
  },
  on: {
    // 🪲 if this transition is triggered, the delayed transition to Fetching never occurs
    // #root.Mode.Demo is a parallel state to #root.Operation.Waiting
    TOGGLE_MODE: {
      target: "#root.Mode.Demo"
    }
  }
}
```

### Expected result

The machine should transition into the #root.Operation.Fetching state after 2000 ms after first entering the Waiting state.

### Actual result

The delayed transition to #root.Operation.Fetching never occurs.

### Reproduction

https://codesandbox.io/s/crazy-dan-jzmhvr?file=/src/index.js

### Additional context

This is an issue with xstate v5.
Occurs with: xstate@5.0.0-beta.34

**Related:** 

https://github.com/statelyai/xstate/issues/4218
https://github.com/statelyai/xstate/pull/4234

The consideration to make:
Should the machine **reenter** the Waiting state when it performs a transition in the parallel state? Reentering Waiting would mean restarting the delayed transition to Fetching.
I feel the Waiting state should not be reentered, since the target of the transition is in a parallel state #root.Mode.

**Comment by Andarist** [comment]
> Should the machine reenter the Waiting state when it performs a transition in the parallel state? 

It's targeting something outside of its source state so I think it's fair to assume that this should be reentered. It's impossible to draw this transition without crossing the Operation's state "boundary" after all.

> Reentering Waiting would mean restarting the delayed transition to Fetching.

This looks broken. This transition should definitely happen at some point - even if Operation is reentered.

**Comment by rlaffers** [comment]
> It's targeting something outside of its source state so I think it's fair to assume that this should be reentered. It's impossible to draw this transition without crossing the Operation's state "boundary" after all.

I'm not sure about this. Re-entering the Waiting state might re-execute some side effects, or restart an invoked actor. This may not be desirable and we would have no way of preventing this (or do we?). 
I would expect that the re-entering of the Waiting state occurs only if it was specifically targeted, i.e.:

```javascript
TOGGLE_MODE: {
    target: ["#root.Mode.Demo", "Waiting"]
}
```

The current behavior becomes even more inappropriate if the Waiting state was not the initial state of the Operation state. For example:

```
Operation
  - Idle  // <-- initial
  - Waiting   // <-- a transition to #root.Mode.Demo occurs
  - Fetching
Mode
  - Normal
  - Demo
```
In the above case, after re-entering Operation, the machine ends up in the `{ Operation: 'Idle', Mode: 'Demo' }` state, whereas the obvious intended state was `{ Operation: 'Waiting', Mode: 'Demo' }`.




> 
> > Reentering Waiting would mean restarting the delayed transition to Fetching.
> 
> This looks broken. This transition should definitely happen at some point - even if Operation is reentered.

Actually, if after re-entering Operation the machine does not enter the Waiting sub-state, then no delayed transition should occur.
And if it did enter re-enter the Waiting state, I don't event think restarting the timer for the delayed transition is desirable. My intuition would require the original timer to proceed uninterupted when I am merely targeting a parallel state.

**To provide a bit more context**: The above snippet is taken from a state machine powering an auto-suggest feature for a text box. User types in an input (the machine goes into Waiting, because we need to debounce fetching the results), and after 500 ms or so the machine then goes Fetching the query results.
The parallel state is for tracking whether user pressed ENTER while in the Waiting state. If he did then handling of the fetched results is slightly different. But right now, if user pressed ENTER while in the Waiting state the machine goes to Idle, and that is a bug because the query results are never fetched.
![auto-suggest](https://github.com/statelyai/xstate/assets/489018/f9382bdb-3020-437b-9354-24188b1b7359)

**Comment by rlaffers** [comment]
I realize, that in my case refactoring this logic is possible (prevent the transition in the parallel state by putting this information into a context, or perhaps another sibling sub-state at the expense of some code duplication, or by moving the transition declaration into the Mode state) - but this may not be always possible, and makes migration to v5 much more involved (the above code worked as intended in v4).

**Comment by rlaffers** [comment]
> > Should the machine reenter the Waiting state when it performs a transition in the parallel state?
> 
> It's targeting something outside of its source state so I think it's fair to assume that this should be reentered. It's impossible to draw this transition without crossing the Operation's state "boundary" after all.

One can draw a line between parallel states without crossing boundaries - we only need to add another dimension. If you think about each orthogonal states as lying in its own plane, than the transition looks like this:

![20231018_082534](https://github.com/statelyai/xstate/assets/489018/38b7dee7-b766-4908-80bd-cc2e6a1819e9)

The green line is a transition targeting a parallel state. The blue line targets a state in the current plane. There is a distinction between these transitions. I am not sure if we need to affirm that unless you can draw a line in 2D space, the source node needs to be exited. If parallel states are really considered orthogonal, thinking about them as two different planes may be a better mental model.

**Comment by Andarist** [comment]
I see your point but the original intention was always to be compatible with SCXML semantics. So the v4 behavior could be considered a bug. Currently, we don't intend to be 100% compatible with SCXML but we still stay inspired by it and other prior work. So, this could be changed and we could diverge from SCXML semantics related to this. We are not sure if we *should* do it here - we'd need a more thorough consideration of this. From one angle - it feels to make sense, from another angle - it introduces more rules to the underlying algorithm and blurs the whole thing. For the time being, we decided not to touch this. That said - I fixed the obvious reentrancy problem here: https://github.com/statelyai/xstate/pull/4368

To coordinate parallel regions you can:
- move this transition to the parallel state and combine it with `stateIn` guard
- `raise` an event in your `Waiting` state and "respond" with a transition in`Mode` by targeting `Demo` there

**Comment by rlaffers** [comment]
That's a fair point. Having just finished reading the SCXML spec, I can see it calls for putting all descendants of the LCCA into the exit set. So in case of parallel regions, this would include the source state.
Even though this significantly limits the practicality of using parallel states (targeting a parallel region will reset the source region which may not be desirable), we can work around this as you pointed out with your examples.

Thanks for fixing the issue with delayed transitions.

Just to clarify: after you fix at #4368, are all entry actions and invocations in the **source** region going to be re-executed/re-invoked?

For example:
```
// S1 and S2 are parallel regions
S1 (onentry=action1, invoke=actor1)
  - S11 (onentry=action11, invoke=actor11)
S2
  - S21
  - S22
```
When a transition in **S11** targets **S22**, both S1 and S2 will be re-entered, and so action1 and action11 are executed again, and both actor1 and actor11 are re-invoked. Correct?

**Comment by Andarist** [comment]
> I can see it calls for putting all descendants of the LCCA into the exit set.

Just so you know, we are using LCA instead of LCCA. With LCCA transitions that have parallel states as their sources always reenter those parallel states and I find this quite surprising - especially with `reenter: false` (the default).

> Just to clarify: after your fix at https://github.com/statelyai/xstate/pull/4368, are all entry actions and invocations in the source region going to be re-executed/re-invoked?

Yes.
